oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java

```diff
@@ -497,6 +501,38 @@ public SampleFamily serviceRelation(DetectPoint detectPoint, List<String> source
         return createMeterSamples(new ServiceRelationEntityDescription(sourceServiceKeys, destServiceKeys, detectPoint, layer, Const.POINT));
     }
 
+    public SampleFamily forEach(List<String> array, Closure<Void> each) {
```

Comment:
@hanahmily @kezhenxu94 @wankai123 Could you recheck here?

Array is a static list in MAL.

Reply:
It's safe to iterate a static array concurrently. But `foreach` is a rare operation in streaming/functional style language. How about picking up [map](https://en.wikipedia.org/wiki/Map_(higher-order_function)) instead? We could pass each sample into the closure to update both values and labels. That seems a more resilient function.

Reply:
`map` is ok to me. But we need to discuss some details. If we use the closure to update the value, we can add `sample` to the closure as the argument which has `tags` and `value` fields, and we can update them in the closure. for the iterate a static array, we can add it in the closure. here is a demo, WDYT?
```yaml
expPrefix: |-
  map({ sample ->
    // dynamic update sample value
    sample.value = sample.value * 2
    // iterate a static array
    def lst = ['client', 'server']
    lst.each { prefix ->
      // dynamic read or update the tags
      if (sample.tags[prefix + '_process_id'] != null) {
        return
      }
      if (sample.tags[prefix + '_local'] == 'true') {
        sample.tags[prefix + '_process_id'] = ProcessRegistry.generateVirtualLocalProcess(tags.service, tags.instance)
        return
      }
      sample.tags[prefix + '_process_id'] = ProcessRegistry.generateVirtualRemoteProcess(tags.service, tags.instance, tags[prefix + '_address'])
    }
  })
```

Reply:
This kind of map seems not better than foreach. 
map usually means iterating elements for the collection calling this method. such as list.map.